crashtracker/src/lib.rs

```diff
@@ -9,26 +9,24 @@
 //!    associated with a crash, and and collects information about the state of
 //!    the program at crash time.  The signal handler runs under a constrained
 //!    environment where many standard operations are illegal.
-//!    https://man7.org/linux/man-pages/man7/signal-safety.7.html
-//!    In particular, memory allocation, and synchronization such as mutexes are
+//!    <https://man7.org/linux/man-pages/man7/signal-safety.7.html>
+//!    In particular, memory allocation, and synchronization such as mutexes, are
 //!    potentially UB.  The signal handler therefore does as little as possible
-//!    in process, and instead writes data across a pipe to a separate receiver
+//!    in process, and instead writes data across a socket to a separate receiver
 //!    process.
-//!    The signal handler then restores the previous signal handler, and waits
-//!    for the receiver process to exit.  Keeping the crashing process alive
-//!    until the receiver has completed increases the chances that the container
-//!    will survive long enough to upload the report; otherwise, there is a
-//!    chance that the container will be killed when the crashing process dies
-//!    and no telemetry will get out.
+//!    The signal handler then waits for the receiver process to exit in order to reap its exit
+//!    status (otherwise, upon the termination of the crashing process the child will be
+//!    re-parented to PID 1 in the current PID namespace, which can be problematic for some user
+//!    applications) and restores the previous signal handler.
 //!    Once the receiver has completed, the crash-handler returns, allowing the
 //!    previous crash handler (if any) to execute, maintaining the customer
 //!    experience as much as possible.
-//! 2. The receiver process runs in the background, listening on `stdin`, which is connected by a
-//!    pipe to the parent process.  When a crash occurs, the receiver gathers the information from
-//!    the pipe, adds additional data about the system state (e.g. /proc/cpuinfo and /proc/meminfo),
-//!    formats it into a crash report, uploads it to the backend, and then exits. The receiver also
-//!    exits if the pipe is closed without a crash report, to avoid leaving a zombie process if the
-//!    parent exits normally.
+//! 2. The receiver process, which is spawned by the signal handler.  It is connected by an
+//!    anynomous AF_UNIX `socketpair()` to the parent process. When a crash occurs, the receiver
```

Comment:
https://man7.org/linux/man-pages/man2/pipe.2.html Pipe pair also could be an option.

Reply:
Indeed.  I was thinking of sticking to AF_UNIX simply because I need to have a named unix socket for some other use-cases.  I can't think of a scenario in which `socketpair()` would be blocked (e.g., by `seccomp()`) but a pipe pair wouldn't.  I also don't really think the two are all-to-distinct for this specific application, except that a streaming AF_UNIX is owned by the kernel networking code and has some slightly better message passing semantics (which we don't use).

Can you think of a good reason to prefer a pipe over a socketpair here?